Adds revert instruction handling

nivida opened the pull request:
## Description

Fixes #176 #1903 #1707 #2197

This PR does implement revert instruction handling in the web3.js lib.
 
If the ``handleRevert`` options property is set to ``true`` you will get revert reasons on calling of ``Contract`` object methods, ``sendTransaction``, and ``call`` of the ``eth`` module. Those revert reasons will get returned to the user over the error listener (``call`` doesn't have a error event) or the catch handler if you are using the native Promise object.

### Error Interface
``` typescript
interface RevertInstructionError extends Error {
    reason: string;
    signature: string;
}
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run dtslint``` with success and extended the tests and types if necessary.
- [x] I ran ```npm run test:unit``` with success and extended the tests if necessary.
- [x] I ran ```npm run build-all``` and tested the resulting file/'s from  ```dist``` folder in a browser.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [x] I have tested my code on the live network.

nivida committed:
revert reason string handling added

nivida committed:
method.isRevertReasonString simplified

nivida committed:
method.getRevertReason funcDoc added and return values improved

nivida committed:
types extended with RevertInstructionError in core-helpers module

nivida committed:
web3-eth types updated

nivida commented on packages/web3-core-method/src/index.js:
```diff
@@ -526,16 +553,34 @@ var getWallet = function(from, accounts) {
         // CALLBACK function
         var sendTxCallback = function(err, result) {
+            if (isCall && (method.isRevertReasonString(result) && method.abiCoder)) {
+                var reason = method.abiCoder.decodeParameter('string', '0x' + result.substring(10));
```
It could be done with the utility functions web3.js does have. But there will be the possibility in the future to have custom error signatures and the AbiCoder would then anyways be used here.

nivida committed:
error handling in Method object of the web3-core-method module fixed

nivida committed:
Method.isRevertReasonString updated

nivida commented on packages/web3-core-method/src/index.js:
```diff
@@ -683,6 +725,56 @@ Method.prototype.buildCall = function() {
+ * @returns {Boolean}
+ */
+Method.prototype.isRevertReasonString = function (data) {
+    return _.isString(data) && ((data.length - 2) / 2) % 32 === 4 && data.substring(0, 10) === '0x08c379a0';
```
```
length = (data.length - 2); // exclude prefix
length = length / 2; // Transforming length to bytes length
argsCount = ((length % 32) === 4 ) // We have 4 arguments: signature, offset, length, data
isErrorSignature = data.substring(0, 10) === '0x08c379a0'; // The first 4 bytes are the Error signature

isRevertReasonString = (argsCount && isErrorSignature)
```

cgewecke reviewed (commented):
Hi, just leaving a quick note. This looks really cool! 

The are couple things that seem like breaking changes
+ Throwing on eth_call
+ Throwing with a different error message on eth_send

Additionally, I'm a little worried about additional server calls to fetch the reason since many wrappers or libs around Web3 are already doing that. In a contract testing context with lots error cases being run it could slow things down to duplicate this logic. 

What do you think about making reason fetching available as an option that is toggled off by default and on via a new contract option? 

Super interested in how `eth_call` works as well...do you think it might be possible to spoof a 'revert' message for a call? Like if data can be set by a contract's users, they could set it so that on 'read', web3 would decode it as an error message and throw?

nivida commented:
@cgewecke 
> This looks really cool!

Thanks!:)

> What do you think about making reason fetching available as an option that is toggled off by default and on via a new contract option?

Sounds good to me but we have to be careful to not have too many configuration options because it could have a bad impact on the DX in the long run.  I would also not have this only as a ``Contract``option instead would I prefer to have this as a "global" ``web3-eth`` module option with the possibility to disable and enable the revert instruction handling for single ``Contract`` objects. This because of ``eth.sendTransaction``, ``eth.sendSignedTransaction``, and ``eth.call`` could also be used to call ``Contract`` functions with the help of the ``eth.abi`` module.

>  Super interested in how eth_call works as well...do you think it might be possible to spoof a 'revert' message for a call? Like if data can be set by a contract's users, they could set it so that on 'read', web3 would decode it as an error message and throw?

I'm not sure if I understood you correctly. But you would like to have the revert instruction handling also for the bare ``eth`` JSON-RPC methods as ``eth.call(...)``? 
This would already be implemented for  ``eth.sendTransaction``, ``eth.sendSignedTransaction``, and ``eth.call``.

nivida committed:
Merge branch '1.x' into feature/revert-reason-string

nivida committed:
console.log added to test case to check the solely remotely failing test case

nivida committed:
handleRevert module option added to eth and eth-contract. Duplicated error removed in Method object

nivida committed:
Basic contract updated, handleRevert option handling fixed, and e2e.method.send extended with severall revert test cases

nivida committed:
revert instruction test cases added to e2e.method.call and assertions in e2e.method.send improved

nivida committed:
eth.handleRevert test added

nivida committed:
Inline comments added to the Method object because of the code complexity

nivida committed:
web3-eth and web3-eth-contract documentation extended with 'handleRevert'

nivida committed:
CHANGELOG.md updated

nivida commented:
@cgewecke I have added all the missing test cases and improved the solution I had with the ``handleRevert`` module option which does default to ``false``. I'm not sure if this module option name is the most explicit one we can have, so feel free to propose a better name! ;) 

During testing of the revert instruction handling did I notice that only module options which are defined before the ``Contract`` instance got created do get applied and all others not. This behavior is existing since the beginning of 1.0 and is actually not that easy to fix as I thought first. 
That’s why I will check that closer later and will open a related issue with a possible solution after.

Example: 
``` javascript
web3.eth.handleRevert = true;
contract = new web3.eth.Contract(..);
contract.methods.method(...).send(...);  // works as expected

-----

contract = new web3.eth.Contract(..); 
contract.handleRevert = true;
contract.methods.method(...).send(...);  // doesn't work as expected
```

Edit:

I've added all new properties to the contract options with commit https://github.com/ethereum/web3.js/pull/3248/commits/9db572c4560f9780abdd4f2f0e84065412794cfd.

Example:

``` javascript
contract = new web3.eth.Contract([], '0x...', {handleRevert: true, ...}); 
contract.methods.method(...).send(...);  // does work as expected
```

ToDo: 

- [x] Extending of ``Contract`` options tests

nivida committed:
Contract options extended to be able to configure singles contracts without re-loading of the object methods

cgewecke commented on packages/web3-core-helpers/src/errors.js:
```diff
@@ -43,5 +43,12 @@ module.exports = {
         return new Error('CONNECTION TIMEOUT: timeout of ' + ms + ' ms achived');
+    },
+    RevertInstructionError: function(reason, signature) {
+        var error = new Error('Your request got reverted with the following reason string: ' + reason);
```
I wonder if it's possible to have this message resemble the existing transaction message more... it looks like this:
```javascript
new Error('Transaction has been reverted by the EVM:\n' + receiptJSON),
```

Lots of people grep that and parse the receipt out of the string as well. The message I propose below is  awkwardly written, but it might provide the same information at a lower cost to people who already have error handling logic in place.

```javascript
'Request errored with revert reason: ' + reason + '. Transaction has been reverted by the EVM:\n' + receiptJSON'
```

And maybe to handle `.call` you would have the function flow be
```javascript
RevertInstructionError: function(reason, signature, receipt){
  if (receipt !== undefined){
    // it's a transaction
 } else {
    // Message is only: 'Request errored with revert reason: ' + reason.'
 }
}
```

nivida replied:
> Lots of people grep that and parse the receipt out of the string as well.

Because the new error message only appears if the ``handleRevert`` property is set to ``true`` do I think can we keep the error message as it is.  

> And maybe to handle .call you would have the function flow be

I thought to keep the logic to only return the receipt for the error listener and to not duplicate the return values with adding it to the ``Error`` object. If the mentioned behavior above is the wished behavior do I propose because of consistency reasons to add the ``receipt`` to all ``Error`` objects.

cgewecke replied:
> Because the new error message only appears if the handleRevert property is set to true do I think can we keep the error message as it is.

What is the advantage of your design?  

>  add the receipt to all Error objects

Yes that seems reasonable to me. 

About serializing the receipt in the message - that decision was made a long time ago and people have adapted to it. They've written all kinds of logic to manage it. Suddenly deciding not to support that interface seems weird to me, idk.

nivida replied:
> What is the advantage of your design?

It doesn't break existing code and will only break it if you use the newly provided property which was, in the beginning, the reason to add it. 

> Yes that seems reasonable to me.

Great I will open a separate PR for this!

> About serializing the receipt in the message - that decision was made a long time ago and people have adapted to it. They've written all kinds of logic to manage it. Suddenly deciding not to support that interface seems weird to me, idk.

I've used another error message to only have one new ``Error`` which is used on ``eth_call`` and ``eth_sendTransaction``. If the serialized receipt is a must-have for those error messages can we also have two different ``Error``'s. This means I would use the old one (``Transaction has been reverted by the EVM:\n' + receiptJSON'``) for ``eth_sendTransaction`` calls and the new one for direct ``eth_call`` calls.

nivida replied:
This would probably allow us to remove the ``handleRevert`` options property at all because the ``Error`` is still the same and does only have additional properties if a revert message is existing.

cgewecke replied:
> I've used another error message to only have one new Error which is used on eth_call and eth_sendTransaction. If the serialized receipt is a must-have for those error messages can we also have two different Error's. 

Wait I don't understand that part. Keeping the error formats similar doesn't seem like it causes lots of problems. 

> This would probably allow us to remove the handleRevert options property at all because the Error is still the same and does only have additional properties if a revert message is existing.

I don't think this true because making additional calls that are already duplicated upstream for this lib potentially slows things down, so it should be optional. There's also completely different behavior for `.call` and potentially some security considerations for Solidity writers.

nivida replied:
I have kept the ``handleRevert`` property as it is and I've added the ``TransactionRevertInstructionError``. This error does have the same error message as we currently do have. With the difference of two additional added object properties called ``reason`` and ``signature``.

cgewecke replied:
I'm unresolving this to look at it again tomorrow, sorry.

nivida replied:
> some security considerations for Solidity writers.

I can't really see what security considerations are here because someone could also call the Ethereum blockchain with his own node and scripts and would get those details returned. There is nothing I think web3.js has to do in this direction. 

> Keeping the error formats similar doesn't seem like it causes lots of problems.

I've chosen a generic error message because the same error was used for ``eth_call`` and ``eth_sendTransaction``. I have created now, in addition, a second error especially for the transaction case which does have the same properties as the generic one but with the small difference of having the same error message as we have currently in 1.2.4. (``'Transaction has been reverted by the EVM:\n' + receiptJSON``)

cgewecke replied:
I think it makes sense to be **as consistent as possible with the existing implementation**. The library has many users who want to see the reason. They'd also like to get it while making the smallest number changes to their own code. That means:

+ New error messages should include the reason. (If this isn't done, there will immediately be issues opened saying feature doesn't work).
+ New error messages should resemble the old message. (So existing handlers are not broken.)
+ New error message for `call` and `send` should resemble each other as much a possible. (So, similar error handlers can be written/used. DRY.) 

I think the above conditions are difficult to get perfect but atm the logic in this PR does not satisfy these requirements.

> I can't really see what security considerations are here because someone could also call the Ethereum....There is nothing I think web3.js has to do in this direction.

Ok, I'm not sure how easy it is to do the fake error I suggested below anyway. But perhaps we could get more opinions about the security implications. I think Web3 *should* worry a little about what it's like to use the library without understanding all the small details:

+ Most users assume that error messages are 'official' and believe them. 
+ Many engineers would be quite surprised if their users could modify the contract such that ordinary data was interpreted by the JS API as an error with a custom message the contract writers had not approved.

nivida replied:
All points above are implemented as described with the last commit. The error message is still the same I've only extended the object property list with ``signature`` and ``reason`` for the ``eth_sendTransaction`` call. The only difference and also only if ``handleRevert`` is set to true is that ``eth_call`` does throw an error if a revert instruction got detected.

 > Many engineers would be quite surprised if their users could modify the contract such that ordinary data was interpreted by the JS API as an error with a custom message the contract writers had not approved.

This is something we never can forbid in JS. The only way to stop a user to modify a contract object would be if web3.js would only run in a JavaScript runtime which does have SES and the related Realms object implemented. The other case you mentioned with spoofing of the requests and returning of wrong errors (malicious node) is also something we only could partially taking action against with having of simultaneously running providers which are connected to different nodes and we compare the returned result of them. A normal middle man attack can get worked against completely by our users using SSL encrypted communication between the client and the connected node.

cgewecke replied:
> All points above are implemented as described with the last commit. The error message is still the same I've only extended the object property list with signature and reason for the eth_sendTransaction call.

Here are the points again:

+  New error messages should include the reason. (If this isn't done, there will immediately be issues opened saying feature doesn't work).
+ New error messages should resemble the old message. (So existing handlers are not broken.)
+ New error message for `call` and `send` should resemble each other as much a possible. (So, similar error handlers can be written/used. DRY.) 

He are the ways it's currently different from the points:
1. The TransactionRevertInstructionError does not include the reason in the error message string. 
2. (Only making this point so that when 1 is done, 2 is also satisfied.)
3. `call` message contains the reason in the error message string, `send` does not. 

Also the message for `call` is this:
```
Your request got reverted with the following reason string:
```

It might make sense to reserve the word 'revert' for state-changing transactions. Most `calls` will just be to read from a view fn. Writing the message so it captures that distinction could be helpful. I would change my original wording for the reason phrase to:
```
Request errored with Solidity reason: ' + reason
```

RE: MITM etc, ok fair enough although I think those issues affect all applications - this question is quite specific to expectations about how Solidity behaves that might not be clear to everyone. I'm not suggesting that the call handling be removed - only that this issue is discussed with others to get their opinion about it's severity and to think about how to disclose it to users (if necessary).

nivida replied:
> The TransactionRevertInstructionError does not include the reason in the error message string.

Yes, to not break the existing code. I would keep this as it is and we can inform the user about the interface in the release and our documentation. The proposed error message from you would be a really ugly solution which I will not implement.

> call message contains the reason in the error message string, send does not.

Yes, because it is a completely new error and does anyways throw where it wasn't throwing an error before. I think this is totally okay to have it like this. 

> Request errored with Solidity reason: ' + reason

This would exclude Vyper and other contract languages. The word ``reverted`` is also correct because the revert instruction got executed in the EVM.

cgewecke replied:
Good point about Vyper.  I also see what you mean about the opcode as well - there's a case there. 

> The proposed error message from you would be a really ugly solution which I will not implement.

Ok. This seems like the root of the disagreement. This discussion has been a waste of everyone's time.

nivida replied:
> This discussion has been a waste of everyone's time.

Yes, the learnings I take away from this is that also changing of an error message string does count as breaking change and that such changes in the "heart" of web3.js should be discussed in detail before the implementation does start. This will probably save time on both sides the proposer and reviewer of a PR. Thanks for your inputs and the review at all!

cgewecke commented on packages/web3-core-method/src/index.js:
```diff
@@ -408,14 +410,42 @@ Method.prototype._confirmTransaction = function(defer, result, payload) {
+                                    );
+                                }
                             } else {
+                                // Throw OOG if status is true and provided gas and used gas are equal
```
Wait! This only a comment but that change was reverted in 1.2.4. If status is true the tx succeeded and shouldn't throw. This gas logic should only run if there's no `status`.

nivida replied:
Oh, sorry this is the logic of 2.x. I will update this inline comment.

cgewecke commented on packages/web3-eth-contract/src/index.js:
```diff
@@ -180,12 +180,13 @@ var Contract = function Contract(jsonInterface, address, options) {
+    this.defaultChain = this.options.defaultChain || this.constructor.defaultChain;
+    this.defaultHardfork = this.options.defaultHardfork || this.constructor.defaultHardfork;
+    this.defaultCommon = this.options.defaultCommon || this.constructor.defaultCommon;
+    this.handleRevert = this.options.handleRevert || this.constructor.handleRevert;
```
Is this one a little tricky because it's boolean? For example if I set 
```javascript
web3.eth.handleRevert = true;
```
and later I change my mind for a specific contract
```javascript
web3.eth.Contract.handleRevert = false
```
I think it will still be `true`.

nivida replied:
The example from you will work and is implemented in the same way [as all other properties](https://github.com/ethereum/web3.js/blob/feature/revert-reason-string/packages/web3-eth/src/index.js#L303). The difference to the ``defaultAccount`` property, for example, is that we are able to define those options over the passed contract constructor options parameter. 

Edit:
Also the ``defaultAccount`` can get defined with the passed options object (``from``).

cgewecke replied:
Oh! Thanks for linking to that part of the code - I misunderstood this.

cgewecke reviewed (commented):
Yes this is looking great @nivida! Just left a couple comments about the err message format and the option behavior. 

Earlier I asked a question about spoofing the revert message with `.call`. With a transaction, the `status` field is checked which means that the tx definitely errored. But with a call, you can only make a guess about what the returned data means. 

I'm wondering if it's possible for a user to set some string value in the contract so that when another person tries to fetch the value, Web3 interprets it as an error and throws?  If so, then it's possible to try to trick someone. 

For example, suppose there is method like:
```solidity
function setNameOfCat(string oldName, string newName) public onlyCatOwner {
  require(isCatOwner(msg.sender, oldName));
  ...etc
}

// Anyone is allowed to read this...
function getCatNameByAddress(address _address) public view returns (string){
  return cats[_address];
}
```

Suppose I am a cat owner and instead of a cat name I set a message with the error prefix that said:
```
In order to read data from this contract you must set an approval for address '0xabc....'  etc...
```

It might be dangerous if someone read my cat's name. Does that make sense? 

**EDIT: Notes on whether this is possible.**

~~Fwiw, I've tried generate a case for this and it's not obvious how it could be done. The `bytes` and `strings` formats returned by the client are very different than the error message format.~~

~~Actually it's possible with `bytes32[]`....~~

Nope, the length check in the reason string evaluator look like it prevents this hack from working.

nivida commented:
> It might be dangerous if someone read my cat's name. Does that make sense?

This is true. But I think it should be handled correctly in the contract or from our consumer with validating the input values. They are able then to show their users the correct error message.

cgewecke commented on packages/web3-core-method/src/index.js:
```diff
@@ -408,14 +410,42 @@ Method.prototype._confirmTransaction = function(defer, result, payload) {
+                                                errors.RevertInstructionError(revertMessage.reason, revertMessage.signature),
+                                                defer.eventEmitter,
+                                                defer.reject,
+                                                payload.callback,
```
Ah yes, what's in this callback? Why is that param different than the existing handler which uses `null`?

nivida replied:
Oh, thanks for seeing this! Yes, the callback should not get passed here because only the ``transactionHash`` does get returned if a callback is existing.

cgewecke commented:
@nivida Could this also be rebased against 1.x at some point so it's possible to see it run against the Truffle E2E?

nivida committed:
Merge branch '1.x' into feature/revert-reason-string

nivida committed:
'eth_sendRawTransaction' does get ignored for the revert handling, inline comment updated, and passed parameters for _fireError in Method object updated

nivida committed:
Merge branch 'feature/revert-reason-string' of github.com:ethereum/web3.js into feature/revert-reason-string

nivida committed:
TransactionRevertInstructionError added and related types updated

nivida committed:
logic checked if the transaction does get signed locally on call of 'eth.sendTransaction' and eth types tests updated

nivida committed:
contract options handling improved, contract test extended, and related documentation updated

nivida committed:
web3-eth.rst updated

cgewecke reviewed (commented):
Apart from the fight about the error message, this LGTM.

WDYT about including test cases for? 
```
web3.eth.call()
web3.eth.sendTransaction()
```

nivida commented:
> WDYT about including test cases for? 

The test cases would be the same and would also test the same code. But I definitely can create a generic test helper for this and extend the test cases for ``eth.call`` and ``eth.sendTransaction``. 👍

cgewecke reviewed (approved):
In summary, this review ends in approval from me, per the PR review guidelines proposed in #3224. AFAIK it doesn't violate any of the requirements. 

(There's are a few dozen unnecessary lints that go against the idea of making the smallest number of changes required, but the core changes seem well-restricted in scope).

nivida committed:
hint about the 'reason' and 'signature' property of the returnd error object added to the documentation

cgewecke commented:
@nivida Final note! This is really cool,  Web3 might be the first lib to collect the revert string from `view` functions and will be very useful. 

Looked pretty safe to me as well - the length of the reason message is hard to replicate using conventional datatypes.

knaderi commented:
Hi,
I upgraded from web3 1.2.4 to 1.2.7.
I have this call in my code:
const transactionReceipt:TransactionReceipt = await web3.eth.sendSignedTransaction(answer.rawTransaction)

error TS2322: Type 'TransactionRevertInstructionError | TransactionReceipt' is not assignable to type 'TransactionReceipt'.
  Type 'TransactionRevertInstructionError' is missing the following properties from type 'TransactionReceipt': status, transactionHash, transactionIndex, blockHash, and 7 more.

57       const transactionReceipt:TransactionReceipt = await web3.eth.sendSignedTransaction(answer.rawTransaction)

cgewecke commented:
@knaderi 

It looks like many of the promise types in web3-eth have mis-defined optional error types. Am opening an issue to correct this right now.

If you *don't* specify the type of your variable (e.g. just let TS infer it) does everything work ok?
```ts
 const transactionReceipt = await web3.eth.sendSignedTransaction(answer.rawTransaction)
```
